### ISSUE TITLE
Best practice for optimizing loss functions in pyro

### ISSUE THREAD

**Issue by ae-foster** [ae-foster/pyro#19]
I'm currently optimizing my loss functions using `opt_eig_ape_loss`:

```
def opt_eig_ape_loss(design, loss_fn, num_samples, num_steps, optim, return_history=False,
                     final_design=None, final_num_samples=None):

    if final_design is None:
        final_design = design
    if final_num_samples is None:
        final_num_samples = num_samples

    params = None
    history = []
    for step in range(num_steps):
        if params is not None:
            pyro.infer.util.zero_grads(params)
        with poutine.trace(param_only=True) as param_capture:
            agg_loss, loss = loss_fn(design, num_samples)
        params = set(site["value"].unconstrained()
                     for site in param_capture.trace.nodes.values())
        if torch.isnan(agg_loss):
            raise ArithmeticError("Encountered NaN loss in opt_eig_ape_loss")
        agg_loss.backward()
        if return_history:
            history.append(loss)
        optim(params)

    _, loss = loss_fn(final_design, final_num_samples, evaluation=True)
    if return_history:
        return torch.stack(history), loss
    else:
        return loss
```

This seems like something that might be implemented in core Pyro, or could be done better using more Pyro tools?

**Comment by ae-foster** [comment]
this is fine
